pl-faded-parsons.py

```diff
@@ -1,314 +1,579 @@
-        if len(distractors) == 0:
-            break
-        index = random.randint(len(distractors))
-        scrambled.append(distractors.pop(index))
-
-    return scrambled, given
-
-
-def base64_encode(s):
-    return base64.b64encode(s.encode("ascii")).decode("ascii")
-
-
-def render_question_panel(element_html, data):
-    """Render the panel that displays the question (from code_lines.txt) and interaction boxes"""
-    element = xml.fragment_fromstring(element_html)
-    answers_name = get_answers_name(element_html)
-
-    format = pl.get_string_attrib(element, "format", "right").replace("-", '_')
-    if format not in ("bottom", "right", "no_code"):
-        raise Exception(f"Unsupported pl-faded-parsons format: \"{format}\". Please see documentation for supported formats")
-
-    lang = pl.get_string_attrib(element, "language", None)
-
-    html_params = {
-        "code_lines": str(element.text),
-    }
+        )
+
+    # this is the `List` in `List[int]`, is None if just `list`
+    wanted_type = get_origin(t)
+    if wanted_type == None:
+        # `t` is a class that's not a dataclass with no annotations, cast it
+        return t(value)
+
+    # this is the `(int,)` in `List[int]`, is `tuple()` if just `list`/`List`
+    type_args = get_args(t)
+    if len(type_args) == 0:
+        return wanted_type(value)
+
+    if wanted_type == list and isinstance(value, list):
+        # `List`/`list` only accepts one type argument
+        item_type = type_args[0]
+        return list(validate_and_instantiate(item_type, v) for v in value)
+
+    if wanted_type == tuple and isinstance(value, tuple):
+        # `Tuple`/`tuple` requires a type argument for each position
+        return tuple(validate_and_instantiate(tt, v) for tt, v in zip(type_args, value))
+
+    if wanted_type == dict and isinstance(value, dict):
+        # `Dict`/`dict` requires 2 type arguments: one for keys, another for values
+        k_type, v_type = type_args
+        return {
+            validate_and_instantiate(k_type, k): validate_and_instantiate(v_type, v)
+            for k, v in value.items()
+        }
 
-    def get_child_text_by_tag(element, tag: str) -> str:
-        """get the innerHTML of the first child of `element` that has the tag `tag`
-        default value is empty string"""
-        return next((elem.text for elem in element if elem.tag == tag), "")
+    raise ParsingError(f"Unhandled case! Could not parse type:{t}, value:{value}")
+
+
+def interleave(list1: list, list2: list) -> list:
+    out = []
+
+    while len(list1) > 0 or len(list2) > 0:
+        if len(list1) > 0:
+            out.append(list1.pop(0))
+        if len(list2) > 0:
+            out.append(list2.pop(0))
+
+    return out
+
+
+def submission_line_to_code(sub_line: Submission.Line) -> str:
+    prefix = sub_line.indent * "    "
+    code = prefix + "".join(interleave(sub_line.codeSnippets, sub_line.blankValues))
+    return code
+
+
+def submission_line_to_mustache(
```

Comment:
Should these be methods on Submission.Line?

Reply:
Ah. no -- the goal is to (after the refactor) eliminate the need for this method entirely. This is kept as a separate method to draw attention as an antipattern. 

see the docstring for this method: "TODO: This routine is technical debt that should be cleaned into FadedParsons class"